### PR TITLE
File and path completion again

### LINKED ISSUE #780: Tab completion behavior for click.File and click.Path types
Lets say this is related to #241 ...

Recently support was added to have argument autocompletion, custom completion functions/lists and a special case for click.Choice options too.

This is great ! I would like to see better support for filenames and directories, have been looking at the code and wanted to share my thoughts here... as far as I can see the problems with filename/directory are multiple but hopefully addressed with a couple of simple patches.

### Undesirable Autocompletion of paths

In the case that the user is completing an option which takes one or more arguments, and those arguments cannot be completed (i.e. they are just a string or integer type), then the builtin completion mechanism correctly reports no results, however the bash completion glue ([here](https://github.com/pallets/click/blob/master/click/_bashcomplete.py#L22)) specifies *"-o default"*

This parameter to the bash *complete* function tells bash to default to completing paths when the resulting COMPREPLY array is empty, and the side effect is undesirable filenames being completed.

### Filenames reported for directory arguments

Because click does not do anything for filenames or directories itself, it instead falls back on the undesirable *complete -o default* behavior mentioned above for filenames and all arguments which did not supply any custom autocompletion list or function.

This means that filenames are suggested when the argument is in fact a directory, and also directory names are suggested for filenames.

### Filename and directory completion has different behavior

The default shell behavior is to take the reported list in COMPREPLY, and if there is only one element, it will append a space and the next TAB is ready to complete the next token, but this is not the desired behavior when completing paths, where you want to complete only one directory at a time.

This can be addressed however by using the *"complete -o nospace ..."* option, which will just cause the shell to not append any space for a single available completion, allowing the program to continue completing the last completion on the next iteration.

In this case, click can:
* Append a space to the reported completion in the case that only one completion was available
* Append a slash (if none is already there), in the case that the completion is a directory name

### Proposed Approach

Instead of handling click.Choice explicitly in _bashcompletion.py, we should instead have the base ParamType cooperate in the process of completion, so that paths and filenames can have an opportunity to decide if the parameter is entirely complete or could be completed further (i.e. whether a space should be appended to the reported completion, or whether it should remain, as in the case for a filename which is a directory that could be completed further).

click.Choice would implement completions by overriding an abstract method of click.ParamType, click.BOOL could also implement this, because why not.

The new *autocompletion* attribute would be checked before ever consulting the click.ParamType abstract method, so the user can always override the default completions for any parameter.

Also, we should change the completion script to not use *"-o default"* and use *"-o nospace"* instead, ensuring that filenames dont show up by default.

Finally, the click.Path and click.File need to do the work we were previously offloading to the shell, this is the most tricky (or *"meaty"*) part of the patch I guess, hopefully a python implementation using *os.listdir()* and file concatenation would not slow down the completion process too much.

### PR DESCRIPTION
**Pull request by dimbleby**
This is almost entirely @gtristan's work from #782, but updated for changes that have been made to `click` in the meantime. fixes #780

In my experience, failing autocompletion on paths and filenames is by *far* the most annoying part about using otherwise lovely `click` CLIs.  Which is why I have been motivated to put this together!

As at #782, the approach is:

- Add `-o nospace` to the bash completion options, so that spaces are not automatically added to completions.  This is necessary so that nested paths can be completed without constantly having to delete the added space.
- Then completions are expected to add spaces themselves, so update all the code that does this for completions automatically generated by `click`
- Look for type-based completions, and implement this for choices, booleans and - especially! - files and paths

I expect that the possibly-controversial bit is that, as of today, any user-implemented autocompletions won't include spaces; and so the addition of `-o nospace` will cause their CLIs to behave slightly differently.

I suggest that we either:

- Just go with this; it's actually an improvement because now users have the ability to do more sophisticated completions (as we do for files and paths)
- Or, we could change the click code to make sure there's a space after anything that users give us.  That denies them the ability to do cool things, but maintains backwards-compatibility.

I'm happy to implement the second of these if that's what maintainers want - especially if it helps this longstanding fix finally to get over the line!

### CODE REVIEW

**Comment by dimbleby** [comment]
I'm unclear what is going on with the PyPy3 failure above in CI:

- I am unable to reproduce the failure locally, 
- CI had it passing at 4231606, and the fix after that (for Windows testscripts) really looks nowhere near the failure

Perhaps some intermittent failure unrelated to these changes?

File: tests/test_bashcomplete.py [hunk 1]
```diff
@@ -400,10 +404,137 @@ def hsub():
...
+    temp = tempfile.mkdtemp()
+    cwd = os.getcwd()
+
+    if chdir:
+        os.chdir(temp)
+
+    yield temp
+
+    os.chdir(cwd)
+    shutil.rmtree(temp)
```
[the remark this answers is not in the document]
Comment (dimbleby) [reply tests/test_bashcomplete.py]: Thanks.  It's only unit test code so whatever, but pushing this change gives me another chance at the unexpectedly broken CI...

**Comment by JustinTArthur** [comment]
I added a PR against your PR to add variable and tilde expansion, @dimbleby.

**Comment by dimbleby** [comment]
So I see!  Looks good, I'll hit the green button in a moment...

**Comment by schollii** [comment]
Awesome, I love click EXCEPT for that annoying space in paths

**Comment by davidism** [comment]
I've been trying to review this and get it ready for 7.1, and I'm starting to realize just how messy completion is. Documentation for completion systems is hard to find and hard to follow.

Bash doesn't quote completions, so filenames with spaces (or any other values with special characters) are invalid. ZSH does quote, but that means it also quotes the trailing space that this PR is manually adding now. But since ZSH uses a menu for completions, not adding a space makes it impossible to continue completing paths, as pressing tab or arrows scrolls through the menu instead of listing the next directory.

Just like this PR turns on `nospace`, we could also turn off quoting in ZSH, and take over doing that. We've already turned off sorting. [argcomplete has a neat trick](https://github.com/kislyuk/argcomplete/pull/159) that turns off spaces only if the completion ends with certain characters, but I don't know how to apply that to ZSH, it doesn't appear to have the equivalent of Bash's `compopt`. But this just starts to feel silly, we're basically making ourselves (or users) recreate everything that the completion system itself is already supposed to provide. Bash and ZSH both have ways of telling the completion system itself to complete files correctly, for example. ZSH already expands `~`.

The problem is the limited way that Click implements completion right now. There's no reason we just have to return completions. In fact, we already support returning descriptions, presumably we could extend that more. Wouldn't it be cool if Click could indicate to the completion script that it should use other functions that Bash or ZSH provide, not just `_describe`?

I'm trying to figure out a way to add some of the enhancements here like the `completions` method and file completion, without making the ZSH experience worse or backing ourselves into a corner when we do want to refactor the whole thing.

**Comment by schollii** [comment]
> I expect that the possibly-controversial bit is that, as of today, any user-implemented autocompletions won't include spaces; and so the addition of -o nospace will cause their CLIs to behave slightly differently.

How about an env var that determines default behaviour? And/or the bash completion code generated (the code meant to be sourced to get completion hooked into bash) could be slightly different based on some arg given. Eg

```
_FOO_BAR_COMPLETE=source foo-bar 
```
vs
```
_FOO_BAR_COMPLETE=source _FOO_BAR_COMP_AUTO_SPACE=false foo-bar
```

**Comment by davidism** [comment]
That wouldn't help the issue here. Completion is already pretty opaque as it is, adding configuration through more env vars is adding complexity. Users shouldn't have to care about that.

**Comment by schollii** [comment]
@davidism Then how about a setting that can be toggled in the CLI main script code eg in my_click_app.py I could add a line click.set_auto_spaces() to get the old behavior, or just keep current behavior by default but have click.set_no_auto_spaces() for those starting fresh. 

Would also be nice to have an additional option on individual commands to override the global so that commands can be tweaked one at a time, as you can in a bash script, but maybe too difficult in click. At least a global for all-or-none would be better than current situation.

**Comment by davidism** [comment]
No, I don't think adding settings *anywhere* is the right idea. Also, turning on our of discs doesn't make Zsh completion any better, it's broken either way. The whole system needs to be redesigned, not made configurable.

**Comment by schollii** [comment]
Redesigning a whole system just to support space seems like a good way for it to never come to fruition.

**Comment by davidism** [comment]
I'm not redesigning it because of spaces. That's *one* reason it needs to be redesigned. It's not even the only reason I called out in my original writeup. Long term consistency and maintainability are the main reasons. Adding configuration options as you've advocated is not in line with these reasons or Click's general design goals.

**Comment by davidism** [comment]
Thanks for working on this. We're rewriting the whole completion system, if this is still and issue after that we'll need to find a new solution that fits with the new system and doesn't have the issues I outlined above. See #1484 and #1622.
